ldp/nn/graph/llm_call_op.py

```diff
@@ -23,71 +31,76 @@
 
-        handler_config = TransformerHandlerConfig(
-            # configurable
-            lm_config=model_config,
-            batch_size=batch_size,
-            max_wait_interval=max_wait_interval,
-            parallel_mode_config=parallel_mode_config,
-            # constant configuration
-            lm_type=LMType.GENERATION,
-            module_call_fn=AsyncTransformerInterface.model_generate,
-            collate_fn=partial(
-                collate_fn_transformer_left_pad, pad_token_id=pad_token_id
-            ),
-            decollate_fn=decollate_fn_transformer_decoder,
-        )
-        self.model_handler = handler_config.make_async_module()
-        self.model_name = model_config.model
+    @classmethod
+    @abstractmethod
+    def prep_tools_for_tokenizer(cls, tools: Tools | None) -> list[dict] | None:
+        """Prepares tools for tokenization."""
 
-        self.llm_call_kwargs = {"logits_processor": LogitsProcessorList()}
+    @classmethod
+    @abstractmethod
+    def parse_tool_request_message(
+        cls, out_text: str, tools: Tools
+    ) -> ToolRequestMessage:
+        """Parses the output text from a tool request message."""
 
-    @staticmethod
-    def prep_messages_for_tokenizer(xi: Messages) -> list[dict]:
+    @classmethod
+    def prep_messages_for_tokenizer(cls, msgs: Messages) -> list[dict]:
+        """Prepares message history for tokenization."""
         result: list[dict] = []
-        for msg in xi:
-            content = msg.content
-            if isinstance(msg, ToolRequestMessage):
-                assert len(msg.tool_calls) == 1, (
-                    "Support parsing only single tool call for now"
-                )
-                tool_call = msg.tool_calls[0]
-                # TODO: document where this format is coming from. Is this a Huggingface chat template syntax?
-                content_dict = {
-                    "name": tool_call.function.name,
-                    "parameters": tool_call.function.arguments,
-                    "thought": msg.content,
-                }
-                content = json.dumps(content_dict)
-            assert content is not None, "content is None, doesn't make sense"
-
+        for msg in msgs:
+            content = cls.get_message_content(msg)
+            assert content is not None, f"Content should not be None: {msg!r}"
             result.append({"role": msg.role, "content": content})
         return result
 
-    @staticmethod
-    def prep_tools_for_tokenizer(tools: list[Tool] | None) -> list[dict] | None:
-        """Prepare tools for the tokenizer by transforming them into a JSON schema."""
+
+class Llama31Parser(MessageAndToolParser):
+    """Follows the Llama 3.1 syntax.
+
+    See details:
+    https://www.llama.com/docs/model-cards-and-prompt-formats/llama3_1/#-tool-calling-(8b/70b/405b)-
+    """
+
+    supported_templates: ClassVar[set[str]] = {
+        "llama2_chat_template_ori.jinja",
```

Comment:
No one but us is going to know what an Ori template is. Any chance you can rename this file, or add a comment explaining it?

Reply:
Yeah it's on my to-do list to rename this, and there's documentation in `chat_templates/README.md` for all of the chat templates. I'll leave it for a later PR to sort this out.